Fix Part of #6240: Subtopic Editor Drag and Drop Skills E2E Test

U8NWXD opened the pull request:
## Explanation

Add an end-to-end test for the subtopic editor to check that a user can drag-and-drop skills in the following ways:
* From the `Uncategorized Skills` area to a subtopic
* From one subtopic to another
* From a subtopic to the `Uncategorized Skills` area

Note: **This PR includes the addition of a dependency**, [`html-dnd`](https://www.npmjs.com/package/html-dnd) (MIT license), to allow simulation of drag-and-drop actions in protractor tests. The normal drag-and-drop protractor functionality is broken by a [bug in selenium](https://code.google.com/p/selenium/issues/detail?id=3604). Here are some relevant issues discussing this problem in the `angular/protractor` repository:

* https://github.com/angular/protractor/issues/123
* https://github.com/angular/protractor/issues/583
* https://github.com/angular/protractor/issues/4353

One of the solutions referenced in these issues was [this gist](https://gist.github.com/druska/624501b7209a74040175), which I was unable to get working.

Fixes part of #6240.

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

U8NWXD committed:
Install html-dnd for Protractor Drag-Drop Tests

A bug in Selenium (see
https://code.google.com/p/selenium/issues/detail?id=3604) prevents us
from using normal Protractor syntax to simulate drag-and-drop events in
a test. Instead, we can use the html-dnd NPM package to simulate
drag-and-drop events.

Fixes part of #6240.

U8NWXD committed:
Add E2E Test for Dragging Skills To Subtopics

In the topic editor, the user can drag skills to assign them to
subtopics. This commit adds an end-to-end test to check that this
functionality works. Note that because of a bug in Selenium preventing
the use of the normal protractor drag-and-drop commands, this commit
makes use of the `html-dnd` package.

Fixes part of #6240.

U8NWXD committed:
Test Dragging Skill Between Subtopics

Update existing e2e test to also check that skills can be dragged
between subtopics.

Fixes part of #6240.

U8NWXD committed:
E2E Test Drag Skill to Uncategorize

Update E2E test to drag skill from subtopic to the uncategorized skills
area to uncategorize the skill.

Fixes part of #6240.

U8NWXD committed:
Refactor Skill Drag to Subtopic Test HTML Classes

To simplify the E2E test checking the dragging of skills to, between,
and from subtopics, this commit replaces the `xpath('..')` used to get
parent DOM elements with a new `protractor-test-*` HTML class for
subtopic columns.

Fixes part of #6240.

U8NWXD committed:
Rename Dragging Skills E2E Test

Renaming reflects that test now covers dragging skills between and from
subtopics, not just to them.

Fixes part of #6240.

nithusha21 commented:
@Showtim3 would you like to take the first pass on this PR?

nithusha21 commented:
Thanks for that info regarding the drag and drop. I think Akshay had tried earlier along those lines but couldn't get it to work. @aks681 the PR description might interest you!

Showtim3 commented on core/tests/protractor_utils/TopicEditorPage.js:
```diff
@@ -171,6 +178,46 @@ var TopicEditorPage = function() {
+      by.css('.protractor-test-assigned-skill-card-text'));
+    var toMove = skillNamesElems.get(skillCardIndex);
+    var target = subtopicColumns.get(toSubtopicIndex);
+    browser.executeScript(
```
Hey @U8NWXD This PR is great. Just one small thing, should be add this ```browser.exectueScript``` to a seperate function, something like ```dragAndDrop(var fromElement, var toElement)``` since this is being used in two places and maybe be used at more places in the future. Also it will improve the code readibility.
@nithusha21

U8NWXD replied:
Done in 4878691de

Showtim3 commented on core/tests/protractor_utils/TopicEditorPage.js:
```diff
@@ -17,6 +17,7 @@
  * in Protractor tests.
  */
 
+const dragAndDrop = require('html-dnd').code;
```
maybe we should let it be a ```var``` to follow the convention below. And in the future, we can replace all the ```var'``` at once. @nithusha21 @U8NWXD

U8NWXD replied:
Done in 4878691de

Showtim3 commented:
@U8NWXD also mind the failing travis CI, quite a lot of tests are failing...

U8NWXD commented:
@Showtim3 I'm seeing the test failures on this PR in the current `develop` as well. Can you reproduce the failures? To reproduce:

```console
$ bash scripts/run_e2e_tests.sh --sharding=false --suite=skillEditor
```

In the browser as the test is running, the test tries to load http://localhost:9001/skill_editor/null, which then results in the 404 error reflected in the CI tests.

Test output:

```
F                                                                                                                                                                                                           
  Skill Editor functionality                                                                                                                                                                                
    ✗ should edit description and concept card explanation                                                                                                                                                  
      - Failed: Create skill button takes too long to be clickable                                                                                                                                          
      Wait timed out after 10118ms                                                                                                                                                                          
      - Failed: No element found using locator: By(css selector, .protractor-test-skill-description-field)                                                                                                  
      - Expected $.length = 8 to equal 0.                                                                                                                                                                   
      Expected $[0] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/ckeditor-bootstrapck-1.0.0/skins/bootstrapck/skin.js?t=J5S8 - Failed to load resource: the server responded 
with a status of 404 (Not Found)', timestamp: 1573107502018, type: '' }) to equal undefined.                                                                                                                
      Expected $[1] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/ckeditor-bootstrapck-1.0.0/skins/bootstrapck/editor.css?t=J5S8 - Failed to load resource: the server respond
ed with a status of 404 (Not Found)', timestamp: 1573107502058, type: '' }) to equal undefined.                                                                                                             
      Expected $[2] = Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/ckeditor-4.12.1/plugins/sharedspace/plugin.js?t=J5S8 - Failed to load resource: the server responded with a 
status of 404 (Not Found)', timestamp: 1573107502172, type: '' }) to equal undefined.                                                                                                                       
      Expected $[3] = Entry({ level: SEVERE, message: 'http://localhost:9001/webpack_bundles/vendors~about~admin~classroom~collection_editor~collection_player~community_dashboard~console_errors~561518cc.b
undle.js 119644:24 Uncaught Error: [CKEDITOR.resourceManager.load] Resource name "sha…tor-4.12.1/plugins/sharedspace/plugin.js?t=J5S8".', timestamp: 1573107502230, type: '' }) to equal undefined.         
      Expected $[4] = Entry({ level: SEVERE, message: 'http://localhost:9001/webpack_bundles/vendors~about~admin~classroom~collection_editor~collection_player~community_dashboard~console_errors~561518cc.b
undle.js 119644:24 Uncaught TypeError: Cannot read property 'icons' of null', timestamp: 1573107502621, type: '' }) to equal undefined.                                                                     
      Expected $[5] = Entry({ level: SEVERE, message: 'http://localhost:9001/webpack_bundles/vendors~about~admin~classroom~collection_editor~collection_player~community_dashboard~console_errors~561518cc.b
undle.js 119644:24 Uncaught TypeError: Cannot read property 'icons' of null', timestamp: 1573107503139, type: '' }) to equal undefined.                                                                     
      Expected $[6] = Entry({ level: SEVERE, message: 'http://localhost:9001/webpack_bundles/vendors~about~admin~classroom~collection_editor~collection_player~community_dashboard~console_errors~561518cc.b
undle.js 119644:24 Uncaught TypeError: Cannot read property 'icons' of null', timestamp: 1573107503627, type: '' }) to equal undefined.                                                                     
      Expected $[7] = Entry({ level: SEVERE, message: 'http://localhost:9001/skill_editor/null - Failed to load resource: the server responded with a status of 404 (Not Found)', timestamp: 1573107516207, 
type: '' }) to equal undefined.
```

It appears that this CI tests has been failing in `develop` since 4ef2e1bed1e61040bb4a642801dd3d1d30d692b7 (PR #7886, [failing TravisCI output](https://travis-ci.org/oppia/oppia/jobs/606567697#L1735))

/cc @nithusha21

Showtim3 commented:
@U8NWXD Hey Chris, yup i verified this, you are correct. The tests are failing on develop and beacuse of the /null route exactly as u said. M just curios how did that PR got in, the travis must have been failing for that PR too right? and how did u find out which PR caused this? Just curious...

Showtim3 commented:
Log: 
![Screenshot 2019-11-07 at 3 29 55 PM](https://user-images.githubusercontent.com/30312043/68379213-79875300-0173-11ea-8a14-2111135a30e3.png)

U8NWXD commented:
@Showtim3 on the [page](https://github.com/oppia/oppia/commits/develop) that lists the commits in `develop`, you can see the CI test status is a red `X` for the past 5 days or so. I checked which tests were failing until I found the first one where the skill editor tests failed, which is the commit I mentioned. The commits have their PRs specified in parentheses at the end of the first line of the message.

U8NWXD commented:
I created #7938 to track fixing the test failures

U8NWXD committed:
Merge branch 'develop' into e2e-subtopics-editor-drag-drop

U8NWXD committed:
Create DragDrop Function and Replace const -> var

Two changes:
* To reduce redundant calls to `browser.executeScript(...)`, create a
  dragAndDrop helper function.
* Use `var` instead of `const` for the `html-dnd` code to match file
  conventions.

Fixes part of #6240. Addresses review comments from @Showtim3.

U8NWXD commented:
@Showtim3 PTAL

U8NWXD commented:
CircleCI says the backend test `core.domain.value_generators_domain_test.ValueGeneratorNameTests` failed, but I haven't been able to reproduce the failure locally. It's passed for me 10 times in a row.

kevinlee12 commented:
@U8NWXD, I'd recommend SSH-ing into the container and debugging, here are the instructions: https://circleci.com/docs/2.0/ssh-access-jobs/

U8NWXD commented:
@kevinlee12 I don't think I have the permissions to rerun the tests with SSH enabled, which according to the instructions you linked is a prerequisite for SSH debugging. The `Rerun workflow` menu is grayed-out for me with a message that I need write permissions.

kevinlee12 commented:
Hmm, in that case, I'd recommend setting up CircleCI on your fork and debugging from there: https://github.com/oppia/oppia/wiki/Setup-your-own-CircleCI-instance

U8NWXD committed:
Use Drag and Drop Helper Function

U8NWXD committed:
Make Drag-Drop Tests Compatible with Others

Because test suites are run on the same instance of the server, the
drag-and-drop test runs on the state left by the previous tests. This
commit updates the test accordingly.

U8NWXD commented:
@Showtim3 PTAL

Showtim3 reviewed (approved):
LGTM!

U8NWXD commented:
@kevinlee12 @vojtechjelinek @nithusha21 @aks681 PTAL

vojtechjelinek reviewed (approved):
LGTM from code owner's perspective. Thanks!

kevinlee12 reviewed (approved):
lgtm!

aks681 reviewed (approved):
LGTM.
